_posts/2023-04-03-date-predicates.md

```diff
@@ -0,0 +1,209 @@
+---
+layout: post
+title: "Just the right time date predicates with Iceberg"
+author: "Marius Grama"
+excerpt_separator: <!--more-->
+image: /assets/blog/date-predicates/christian-pfeifer-l6OraG-v0d8-unsplash.jpg
+---
+
+In the data lake world, data partitioning is a technique that is critical
+to the performance of read operations. In order to avoid scanning large 
+amounts of data accidentally, and also to limit the number of partitions 
+that are being processed by a query, a query engine must push down constant 
+expressions when filtering partitions. 
+
+<!--more-->
+
+Partitions in an Iceberg table tend to be fairly large, containing up to 
+tens or even hundreds of data files. It is therefore crucial to be able to 
+skip irrelevant partitions while scanning a table in order to ensure high 
+performance query processing speed. When a table is created in a data lake, 
+its partitioning scheme constitutes a de-facto index, speeding up queries 
+against it by pruning out irrelevant partitions from the scan operation. 
+
+Date and time are natural and universal partitioning candidates. 
+Common partition patterns revolve around month, day, hour. 
+One exciting feature  of the Iceberg table format is its 
+[hidden partitioning](https://trino.io/blog/2021/07/12/in-place-table-evolution-and-cloud-compatibility-with-iceberg.html#partition-specification-evolution).
+Iceberg uses handy [transforms](https://trino.io/docs/current/connector/iceberg.html#partitioned-tables) 
+such as `year`, `month`, `day`, `hour` to deal with the complexities of
+mapping a raw timestamp value to an actual partition value in a manner
+that is transparent to the user. 
+
+Let’s look at a typical example of an Iceberg table containing log events 
+which are partitioned by day:
+
+```sql
+CREATE TABLE logs (
+    event_time timestamp(6) with time zone,
+    level varchar,
+    message varchar) 
+WITH (partitioning=ARRAY['day(event_time)'])
+```
+
+When dealing with logs, it often happens that we want to know what 
+happened today or within the last few days:
+
+```sql
+SELECT * 
+FROM logs 
+WHERE
+      event_time >= CURRENT_DATE
+```
+
+```sql
+SELECT * 
+FROM logs 
+WHERE 
+      event_time >= CURRENT_DATE - INTERVAL '7' DAY
+```
+
+## Constant folding
+
+Trino uses the _constant folding_ optimization technique for dealing 
+with these types of queries by internally rewriting the filter expression 
+as a comparison predicate against a constant evaluated before executing
+the query in order to avoid recalculating the same expression for each
+row scanned:
+
+![](/assets/blog/date-predicates/constant_folding.png)
```

Comment:
There's a slight change needed for this image - to avoid having Word like single quotes around `7`.

Reply:
Done.